src/box/space.c

```diff
@@ -1344,13 +1344,34 @@ after_old_tuple_lookup:;
 
 /**
  * Executes an IPROTO_INSERT_ARROW request. The request contains the data in
- * Arrow columnar format.
+ * Arrow columnar format that can appear in two guises:
+ *  1. in-memory data structures (arrow_array and arrow_schema);
+ *  2. serialized for interprocess communication (arrow_ipc and arrow_ipc_end).
  * Can return nonzero in case of error (diag is set).
  */
 static int
 space_execute_insert_arrow(struct space *space, struct txn *txn,
 			   struct request *request)
 {
+	if (request->arrow_array != NULL) {
+		assert(request->arrow_schema != NULL);
+		assert(request->arrow_ipc == NULL);
+		assert(request->arrow_ipc_end == NULL);
+		struct ArrowArray *array = request->arrow_array;
+		struct ArrowSchema *schema = request->arrow_schema;
+
+		if (space->vtab->execute_insert_arrow(
+				space, txn, array, schema) != 0)
+			return -1;
+		struct region *txn_region = tx_region_acquire(txn);
+		int rc = arrow_ipc_encode(
+				array, schema, txn_region, &request->arrow_ipc,
+				&request->arrow_ipc_end);
+		tx_region_release(txn, TX_ALLOC_SYSTEM);
+		return rc;
+	}
```

Comment:
Can we always decode `request->arrow_ipc` and set `arrow_schema` and `arrow_array` in `xrow_decode_dml`?

Can we set `request->arrow_ipc` by encoding `arrow_array` and `arrow_schema` in `txn_add_redo`?

Reply:
> Can we always decode request->arrow_ipc and set arrow_schema and arrow_array in xrow_decode_dml?

It's hard to release `array` and `schema` if they are allocated from `xrow_decode_dml()`.

> Can we set request->arrow_ipc by encoding arrow_array and arrow_schema in txn_add_redo?

Done.